appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/eesecurity/JwtPublicKeyStore.java

```diff
@@ -0,0 +1,297 @@
+            }
+        }
+    }
+
+    private PublicKey createPublicKeyFromPem(String key) throws Exception {
+        key = key.replaceAll("-----BEGIN (.*)-----", "")
+                .replaceAll("-----END (.*)----", "")
+                .replaceAll("\r\n", "")
+                .replaceAll("\n", "")
+                .trim();
+
+        byte[] keyBytes = Base64.getDecoder().decode(key);
+        X509EncodedKeySpec publicKeySpec = new X509EncodedKeySpec(keyBytes);
+        return KeyFactory.getInstance(RSA_ALGORITHM)
+                .generatePublic(publicKeySpec);
+
+    }
+
+    private PublicKey createPublicKeyFromJWKS(String jwksValue, String keyID) throws Exception {
+        JsonObject jwks = parseJwks(jwksValue);
+        JsonArray keys = jwks.getJsonArray("keys");
+        JsonObject jwk = keys != null ? findJwk(keys, keyID) : jwks;
+
+        // the public exponent
+        byte[] exponentBytes = Base64.getUrlDecoder().decode(jwk.getString("e"));
+        BigInteger exponent = new BigInteger(1, exponentBytes);
+
+        // the modulus
+        byte[] modulusBytes = Base64.getUrlDecoder().decode(jwk.getString("n"));
+        BigInteger modulus = new BigInteger(1, modulusBytes);
+
+        RSAPublicKeySpec publicKeySpec = new RSAPublicKeySpec(modulus, exponent);
+        return KeyFactory.getInstance(RSA_ALGORITHM)
+                .generatePublic(publicKeySpec);
+    }
+
+    private JsonObject parseJwks(String jwksValue) throws Exception {
+        JsonObject jwks;
+        try (JsonReader reader = Json.createReader(new StringReader(jwksValue))) {
+            jwks = reader.readObject();
+        } catch (Exception ex) {
+            // if jwks is encoded
+            byte[] jwksDecodedValue = Base64.getDecoder().decode(jwksValue);
+            try (InputStream jwksStream = new ByteArrayInputStream(jwksDecodedValue);
+                    JsonReader reader = Json.createReader(jwksStream)) {
+                jwks = reader.readObject();
+            }
+        }
+        return jwks;
+    }
+
+    private JsonObject findJwk(JsonArray keys, String keyID) {
+        if (Objects.isNull(keyID) && keys.size() > 0) {
+            return keys.getJsonObject(0);
+        }
+
+        for (JsonValue value : keys) {
+            JsonObject jwk = value.asJsonObject();
+            if (Objects.equals(keyID, jwk.getString("kid"))) {
+                return jwk;
+            }
+        }
+
+        throw new IllegalStateException("No matching JWK for KeyID.");
+    }
+    
+    private static class PublicKeyLoadingCache {
+        
+        private final long ttl;
+        private final Supplier<Optional<String>> loadingFunction;
+        private long lastUpdated;
+        private Optional<String> publicKey;
+        
+        
+        public PublicKeyLoadingCache(long ttl, Supplier<Optional<String>> loadingFunction) {
+            this.ttl = ttl;
+            this.loadingFunction = loadingFunction;
+            this.lastUpdated = 0;
```

Comment:
This statement is redundant since the value has been already initialized to `0`.